### PR TITLE
cilium: improve 16bit ifindex probe on old kernels

### LINKED ISSUE #27642: Improve HAVE_FIB_IFINDEX probe
https://github.com/cilium/cilium/pull/27622 lifted a limitation that requires Cilium-managed native devices (where `bpf_host` is attached) to have an ifindex <= MAX_UINT16.

This limitation no longer applies on systems with a kernel that has the `HAVE_FIB_IFINDEX` feature. Currently this feature is only detected on kernels >= 5.10. But the [relevant patch](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=d1c362e1dd68a421cf9033404cf141a4ab734a5d) was actually backported to older kernels. It would be good to also detect the feature on those older kernels, to further reduce the impact of the MAX_UINT16 limitation.

Quoting @borkmann
>this can be further refined by not relying
on the asm.FnRedirectPeer helper presence but by actually doing a runtime
BPF program probe so that stable kernels can even be covered.

When this is fixed, we should also add `HAVE_FIB_IFINDEX` to the complexity test configs for the relevant kernels.

### PR DESCRIPTION
**Pull request by msune**
Previously, `HaveFibIfindex` probe checked the presence of commit `d1c362e1dd68` in the Linux kernel indirectly, by verifying the presence of the redirect helpers (which were merged alongside).

Since `d1c362e1dd68` has been backported to older kernels - but not helpers -, this commit improves `HaveFibIfindex` detection by running a live probe.

Fixes: #27642

```release-note
On kernels where `d1c362e1dd68` has been backported, don't store `ifindex` in the CT map
```

### CODE REVIEW

**Comment by msune** [comment]
RFC

I don't think the unit is enough. I would think adding a check in integration/e2e that checks the flag is correctly detected would be a better approach than a unit (which I am not even sure is actually correct for backports as is). Please advise

File: pkg/datapath/linux/probes/probes.go [hunk 1]
```diff
@@ -386,10 +387,145 @@ func HaveBoundedLoops() error {
...
+		}
+
+		//Attach eBPF program to call bpf_fib_lookup()
+		probe_name := "have_fib_index_probe"
+		progSpec := &ebpf.ProgramSpec{
+			Type: ebpf.SchedCLS,
+			License: "GPL",
+		}
+		progSpec.Instructions = asm.Instructions{
+			//Keep skbuf ptr (JIC)
```
Comment (lmb) [inline pkg/datapath/linux/probes/probes.go]: What does JIC mean?

File: pkg/datapath/linux/probes/probes.go [hunk 2]
```diff
@@ -386,10 +387,145 @@ func HaveBoundedLoops() error {
...
+	}()
+
+	ns, err := netns.New()
+	if err != nil {
+		return fmt.Errorf("create netns: %w", err)
+	}
+	defer ns.Close()
+
+	return ns.Do(func() error {
+		lo, err := netlink.LinkByName("lo");
```
Comment (lmb) [inline pkg/datapath/linux/probes/probes.go]: How does the `netlink` package figure out which network namespace to use? I'm asking because using the thread local netns is very fiddly in Go and only works if there are no goroutines in play.

File: pkg/datapath/linux/probes/probes.go [hunk 3]
```diff
@@ -386,10 +387,145 @@ func HaveBoundedLoops() error {
...
 
 // HaveFibIfindex checks if kernel has d1c362e1dd68 ("bpf: Always return target
-// ifindex in bpf_fib_lookup") which is 5.10+. This got merged in the same kernel
-// as the new redirect helpers.
-func HaveFibIfindex() error {
-	return features.HaveProgramHelper(ebpf.SchedCLS, asm.FnRedirectPeer)
+// ifindex in bpf_fib_lookup") which is 5.10+ (and backports). It does so by
+// checking that ifindex has been updated as a result of the RIB lookup even
+// though the bpf_fib_lookup failed (no L2 resolution for the NH)
+func HaveFibIfindex() (err error) {
```
Comment (lmb) [inline pkg/datapath/linux/probes/probes.go:394]: How often is this function called? Right now it doesn't do any caching of the result, which features.HaveProgramHelper does.

File: pkg/datapath/linux/probes/probes.go [hunk 4]
```diff
@@ -386,10 +387,145 @@ func HaveBoundedLoops() error {
...
+	defer func() {
+		if err != nil && !errors.Is(err, ebpf.ErrNotSupported) {
+			log.WithError(err).Fatal("failed to probe HaveFibIfindex")
```
Comment (lmb) [inline pkg/datapath/linux/probes/probes.go:397]: Does it make more sense to log in the caller? That would shorten the function body by a lot.

File: pkg/datapath/linux/probes/probes.go [hunk 5]
```diff
@@ -386,10 +387,145 @@ func HaveBoundedLoops() error {
...
+			asm.StoreMem(asm.RFP, -64, asm.R2, asm.DWord),
+			asm.StoreMem(asm.RFP, -72, asm.R2, asm.DWord),
+			asm.StoreMem(asm.RFP, -80, asm.R2, asm.DWord),
+
+			//Set AF_INET
+			asm.Mov.Imm(asm.R2, 2),
+			asm.StoreMem(asm.RFP, -80, asm.R2, asm.DWord),
+
+			//Set dst IP (1.2.3.4)
+			asm.Mov.Imm(asm.R2, 0x04030201),
```
Comment (lmb) [inline pkg/datapath/linux/probes/probes.go]: What about the endianness?

File: pkg/datapath/linux/probes/probes.go [hunk 6]
```diff
@@ -386,10 +387,145 @@ func HaveBoundedLoops() error {
...
+			//Set ifindex to lo (1)
+			asm.Mov.Imm(asm.R2, 1),
+			asm.StoreMem(asm.RFP, -72, asm.R2, asm.DWord),
+
+			//Args
+			asm.Mov.Reg(asm.R1, asm.R8), //skbuff ptr
+			asm.Mov.Reg(asm.R2, asm.RFP), //struct bpf_fib_lookup
+			asm.Add.Imm(asm.R2, -80),
+			asm.Mov.Imm(asm.R3, 64), //len
+			asm.Mov.Imm(asm.R4, 4), //flags: BPF_FIB_LOOKUP_SKIP_NEIGH
```
Comment (lmb) [inline pkg/datapath/linux/probes/probes.go]: Nit: you can probably use `unix.BPF_FIB_LOOKUP...` directly? Also, maybe a comment makes sense indicating why that flag is needed?

File: pkg/datapath/linux/probes/probes_test.go [hunk 7]
```diff
@@ -344,3 +344,16 @@ func TestHaveTCX(t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
+func TestDetectHaveFibIfindex(t *testing.T) {
+	testutils.PrivilegedTest(t)
+	testutils.SkipOnOldKernel(t, "5.10", "bpf: Always return target ifindex in bpf_fib_lookup")
```
Comment (lmb) [inline pkg/datapath/linux/probes/probes_test.go:350]: Why? On old kernels it should still work, but return an error saying that the feature is not supported, no?

File: pkg/datapath/linux/probes/probes.go (hunk 1 above)
Comment (msune) [reply pkg/datapath/linux/probes/probes.go]: Just in case.

We don't need R1 anymore after the call today, so it can be removed. will do.

File: pkg/datapath/linux/probes/probes.go (hunk 4 above)
Comment (msune) [reply pkg/datapath/linux/probes/probes.go:397]: I honestly don't know. I am new to golang (literally my first program).

I've seen the same approach being used in other probes (E.g. https://github.com/cilium/cilium/blob/4733a894097f79d98c9872cc85b3ca07c6a6b188/pkg/datapath/linux/probes/probes.go#L669), and I thought it was kind of a good pattern to have a unified end trace of error in case the probe failed.

Not sure if at the end it will be better (e.g. copying the string on each err. path print)

File: pkg/datapath/linux/probes/probes.go (hunk 3 above)
Comment (msune) [reply pkg/datapath/linux/probes/probes.go:394]: Good point. I am not really sure. I can tell that is being executed 2x during startup, but I don't know if it's being executed during normal operation.

Let me investigate how could it be cached

File: pkg/datapath/linux/probes/probes.go (hunk 6 above)
Comment (msune) [reply pkg/datapath/linux/probes/probes.go]: Can you? :D OK, let me try.

File: pkg/datapath/linux/probes/probes.go (hunk 5 above)
Comment (msune) [reply pkg/datapath/linux/probes/probes.go]: Argh. Good catch. I forgot to fix this in the latest version. Thx.

File: pkg/datapath/linux/probes/probes_test.go (hunk 7 above)
Comment (msune) [reply pkg/datapath/linux/probes/probes_test.go:350]: Humm.. `HaveFibIfindex` call will return `err` = `ebpf.ErrNotSupported`, so no, the test - as written here - shouldn't succeed for Kernels that don't have the patch.

But as I say in the comment, I see little value on this unit, as you don't have much control on which Kernel the CI is actually running. I think a much better approach would be to run it as part of the e2e or integr, where tests are run against specific kernels, making sure detection (true/false) is correct based on the specific kernel version.

File: pkg/datapath/linux/probes/probes.go (hunk 2 above)
Comment (msune) [reply pkg/datapath/linux/probes/probes.go]: As far as I can see it doesn't use co-routines except when subscribing to events - which are not used in this patch.

In fact, this approach is used in many parts of the code from what I can see. Look at one random example:

https://github.com/cilium/cilium/blob/main/plugins/cilium-cni/chaining/generic-veth/generic-veth.go#L63-L78

For your reference:

```
/dev/netlink$ grep -ne "go \w*(" * 
addr_linux.go:351:		go func() {
addr_linux.go:365:	go func() {
grep: cmd: Is a directory
link_linux.go:2418:		go func() {
link_linux.go:2432:	go func() {
neigh_linux.go:403:		go func() {
neigh_linux.go:414:	go func() {
grep: nl: Is a directory
proc_event_linux.go:140:		go func() {
proc_event_linux.go:146:	go func() {
route_linux.go:1641:		go func() {
route_linux.go:1655:	go func() {
grep: testdata: Is a directory
xfrm_monitor_linux.go:48:		go func() {
xfrm_monitor_linux.go:55:	go func() {
```

If you want, I can add a sanity check after the link creation to make sure `ifindex` is indeed `2`, so making sure we are in the test NS. But as said.. if netlink uses go coroutines at some point to do this basic ops, the last of the problems will be this probe :)

File: pkg/datapath/linux/probes/probes.go (hunk 1 above)
Comment (msune) [reply pkg/datapath/linux/probes/probes.go]: Done.

File: pkg/datapath/linux/probes/probes.go (hunk 5 above)
Comment (msune) [reply pkg/datapath/linux/probes/probes.go]: Done, please have a look

File: pkg/datapath/linux/probes/probes.go (hunk 6 above)
Comment (msune) [reply pkg/datapath/linux/probes/probes.go]: Done. Awesome, thx!

File: pkg/datapath/linux/probes/probes.go (hunk 4 above)
Comment (lmb) [reply pkg/datapath/linux/probes/probes.go:397]: No worries! Here is how I think about this:
- The Have* functions are meant to detect whether a feature is present or not. This means that the calling code will adapt to an error being returned. This would bias me towards not logging in the function itself.
- Feature probes are a bit tricky since they can actually have three separate outcomes: supported, not supported, inconclusive (due to some unforeseen error). The code you've linked to only logs the inconclusive case, which makes sense to me.

To sum up: I would remove the logging throughout the function and keep this defer call at the top.

File: pkg/datapath/linux/probes/probes.go (hunk 3 above)
Comment (lmb) [reply pkg/datapath/linux/probes/probes.go:394]: For caching there is https://go.dev/pkg/sync/?m=old#OnceValues If other functions in the package don't cache you probably also don't have to cache here.

File: pkg/datapath/linux/probes/probes.go (hunk 2 above)
Comment (lmb) [reply pkg/datapath/linux/probes/probes.go]: Fine if other parts do this as well, I thought I'd ask since I'm not familiar with our netlink code.

File: pkg/datapath/linux/probes/probes_test.go (hunk 7 above)
Comment (lmb) [reply pkg/datapath/linux/probes/probes_test.go:350]: Good question, I don't actually know how we test these. ci-runtime only runs on whatever CI has.

File: pkg/datapath/linux/probes/probes.go (hunk 4 above)
Comment (lmb) [reply pkg/datapath/linux/probes/probes.go:397]: P.S. Due to how the `defer` works you are actually logging twice. `defer func() {}()` creates a closure which has access to the __return variables___ of the enclosing function. See this example: https://go.dev/play/p/6tiSrxY2_bu

See the language spec: https://go.dev/ref/spec#Defer_statements
> That is, if the surrounding function returns through an explicit [return statement](https://go.dev/ref/spec#Return_statements), deferred functions are executed after any result parameters are set by that return statement but before the function returns to its caller.

File: pkg/datapath/linux/probes/probes.go (hunk 3 above)
Comment (msune) [reply pkg/datapath/linux/probes/probes.go:394]: So first, other runtime probes are not cached (e.g. [HaveOuterSourceIPSupport()](https://github.com/cilium/cilium/blob/14f0780195f498a2395089c56a871caecfd17f61/pkg/datapath/linux/probes/probes.go#L472) or [HaveOuterSourceIPSupport()](https://github.com/cilium/cilium/blob/3f5e735cdb303cdcba019a02f46cc781006dc386/pkg/datapath/linux/probes/probes.go#L532)), albeit they are not called more than once.

The currrent `HaveFibIfindex()` function is called from three different places:

* Directly from  https://github.com/cilium/cilium/blob/3f5e735cdb303cdcba019a02f46cc781006dc386/daemon/cmd/kube_proxy_replacement.go#L404
* From Unix metrics (`enableIfIndexMetric()`): https://github.com/cilium/cilium/blob/3f5e735cdb303cdcba019a02f46cc781006dc386/pkg/metrics/metrics_unix.go#L29
* From `ExecuteHeaderProbes` [here](https://github.com/cilium/cilium/blob/3f5e735cdb303cdcba019a02f46cc781006dc386/pkg/datapath/linux/probes/probes.go#L720), which then is cached in a boolean flag. `ExecuteHeaderProbes` is later called by the `daemon_main.go`: https://github.com/cilium/cilium/blob/3f5e735cdb303cdcba019a02f46cc781006dc386/daemon/cmd/daemon_main.go#L1275. `.

There is probably a reason (not all daemons require all probes?), can't tell right now, but sounds a bit convoluted.

Side comment, wondering if a singleton pattern for _all_ probes would make sense, so that there is a tight control over _when_ probes run and can be easily and cleanly cached cached.

In any case, a) this doesn't seem anything strictly related to this PR b) to my understanding this doesn't run other than at startup.

Let me know if you think I should address this in this PR, open an issue or s/t.

File: pkg/datapath/linux/probes/probes.go (hunk 4 above)
Comment (msune) [reply pkg/datapath/linux/probes/probes.go:397]: >To sum up: I would remove the logging throughout the function and keep this defer call at the top.

Logging on error paths - being golang or not - is nothing new. Error specifics have better context locally, or closer to the error as possible (either through direct logging or returning the error with the string).

I think the _reason why_ a probe has failed is important to be printed as probes in general shall NOT fail. They can evaluate to true or false, but they shall always work. So I very much think somehow the specific error must be printed, somewhere.

I am fine removing the DBG trace and I am also fine moving specific error strings to `err`, so that they can later be dumped in the `defer()`. Deal?

> P.S. Due to how the defer works you are actually logging twice. defer func() {}() creates a closure which has access to the return variables_ of the enclosing function. See this example: https://go.dev/play/p/6tiSrxY2_bu

Yes, I am familiar with `defer()` (other langs have similar constructs), and as I explain above, it was kind of intentional. But I agree it's probably redundant, so I can move the string to `err`.

Comment (lmb) [reply pkg/datapath/linux/probes/probes.go:397]: > I think the _reason why_ a probe has failed is important to be printed as probes in general shall NOT fail.

Yes, the canonical way to do that is `return fmt.Errorf("more context: %w", err)` or similar, not log statements (imo).

Comment (msune) [reply pkg/datapath/linux/probes/probes.go:397]: Done

File: pkg/datapath/linux/probes/probes.go [hunk 8]
```diff
@@ -386,10 +388,137 @@ func HaveBoundedLoops() error {
...
+// checking that ifindex has been updated as a result of the RIB lookup even
+// though the bpf_fib_lookup failed (no L2 resolution for the NH)
+func HaveFibIfindex() (err error) {
+	defer func() {
+		if err != nil && !errors.Is(err, ebpf.ErrNotSupported) {
+			log.WithError(err).Fatal("failed to probe HaveFibIfindex")
+		}
+
+		if errors.Is(err, ebpf.ErrNotSupported) {
+			log.Info("HaveFibIfindex NOT supported")
```
Comment (lmb) [inline pkg/datapath/linux/probes/probes.go]: Nit: make this debug?

**Review by lmb — CHANGES_REQUESTED** [review]
There are a couple of vet errors due to `fmt.Errorf`.

File: pkg/datapath/linux/probes/probes.go (hunk 8 above)
Comment (msune) [reply pkg/datapath/linux/probes/probes.go]: I don't have a strong opinion.

Moving it to `Dbg` together with fixing the errors of `govet` (thx, wasn't aware of that `make` target)

Comment (msune) [reply pkg/datapath/linux/probes/probes.go]: Done

**Review by lmb — APPROVED** [review]
Still a CI failure unfortunately: https://github.com/cilium/cilium/actions/runs/9077690092/job/24943135667?pr=32466#step:14:35

**Comment by lmb** [comment]
You might need a
```
go get golang.org/x/sys/unix@latest
go mod vendor
```

**Comment by julianwiedmann** [comment]
I was discussing with Marc a bit on how this might affect downgrades.

Considering the scenario where a 1.16 Cilium would detect `HAVE_FIB_IFINDEX` on a fixed 5.4 kernel, create new CT entries (without filling the `ifindex`) ... and then we downgrade into 1.15, which doesn't detect `HAVE_FIB_IFINDEX` and relies on the non-existent `ifindex` value. We could solve this by having some compat code in 1.16 that *continues* to fill the ifindex (without using it in the datapath), to enable the downgrade. And then in 1.17 remove that compat code.

But Marc also made the fair point that we didn't consider this downgrade aspect when merging https://github.com/cilium/cilium/pull/27622. Which also produced such CT states, that were not safe for downgrade.

**Comment by lmb** [comment]
Based on my earlier work to bump the [minimum kernel version](https://github.com/cilium/cilium/issues/30456) we know that at least ubuntu 20.04 lts started on a 5.4 kernel. But the current version in repos is 5.10: https://packages.ubuntu.com/search?suite=focal&searchon=names&keywords=linux-image Probably this would only people that hand roll their kernels / distros and are on an old version?

File: pkg/datapath/linux/probes/probes_test.go (hunk 7 above)
Comment (msune) [reply pkg/datapath/linux/probes/probes_test.go:350]: @julianwiedmann?

Comment (julianwiedmann) [reply pkg/datapath/linux/probes/probes_test.go:350]: @lmb I was discussing with Marc a bit more. Should this probe actually live in `cilium/ebpf/features`, and would `cilium/ebpf` have infrastructure for more detailed testing of the probe code, across different kernels etc?

Comment (lmb) [reply pkg/datapath/linux/probes/probes_test.go:350]: Yes, testing of the probe would definitely be more thorough in c/ebpf since we have CI against all supported kernel.org releases. Pulling in @rgo3 since he maintains `features`.

Comment (rgo3) [reply pkg/datapath/linux/probes/probes_test.go:350]: I think the current split of having `c/ebpf/features` and `pkg/datapath/linux/probes` is intentional. `c/ebpf/features` was only intended for more generalized APIs, like the ones we currently have, e.g. `HaveProgramType`, `HaveMapType`,  `HaveProgramHelper`. Specifically with `HaveProgramHelper` there is the issue that it only cares about the presence of a certain helper, callers can't really know which specific functionality this helper offers if it's behavior changed over time in the kernel. IMO if individual projects (including Cilium) need to know about specific behavior those projects should maintain more specific probes themselves. So I think this probe is in the right place, but we might need to improve Cilium's CI to run `ci-runtime` tests against different kernel versions if otherwise the testing of our internal probes doesn't have much value.

File: pkg/datapath/linux/probes/probes.go [hunk 9]
```diff
@@ -386,10 +388,137 @@ func HaveBoundedLoops() error {
...
+			asm.Return(),
+		}
+
+		prog, err := ebpf.NewProgram(progSpec)
+		if err != nil {
+			return fmt.Errorf("unable to load eBPF program '%s': %w", probe_name, err)
+		}
+		defer prog.Close()
+
+		pkt := []byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0}
```
Comment (ti-mo) [inline pkg/datapath/linux/probes/probes.go]: Slightly more expressive:
```suggestion
		pkt := make([]byte, 16)
```

File: pkg/datapath/linux/probes/probes.go [hunk 10]
```diff
@@ -386,10 +388,137 @@ func HaveBoundedLoops() error {
...
+		}
+		if err := netlink.RouteAdd(route); err != nil {
+			return fmt.Errorf("unable to add route %s via %s dev dummy: %w", dstCIDR, gwAddrStr, err)
+		}
+
+		//netlink stores bytes in NBO, so use Native to conver to uint32
+		nhBe := binary.NativeEndian.Uint32(dst.IP.To4())
+
+		//Attach eBPF program to call bpf_fib_lookup()
+		probe_name := "have_fib_index_probe"
```
Comment (ti-mo) [inline pkg/datapath/linux/probes/probes.go:459]: Looks like this isn't actually put into the ProgramSpec, so maybe it doesn't really need to be a variable. Or it should go into the spec.

File: pkg/datapath/linux/probes/probes.go [hunk 11]
```diff
@@ -386,10 +388,137 @@ func HaveBoundedLoops() error {
...
+	}()
+
+	ns, err := netns.New()
+	if err != nil {
+		return fmt.Errorf("create netns: %w", err)
+	}
+	defer ns.Close()
+
+	return ns.Do(func() error {
+		lo, err := netlink.LinkByName("lo")
```
Comment (ti-mo) [inline pkg/datapath/linux/probes/probes.go:412]: Is the loopback interface actually used for something?

Why not attach the program to `lo` instead of creating a dummy?

File: pkg/datapath/linux/probes/probes.go [hunk 12]
```diff
@@ -386,10 +388,137 @@ func HaveBoundedLoops() error {
...
+			asm.Mov.Imm(asm.R3, 64),                             //len
+			asm.Mov.Imm(asm.R4, unix.BPF_FIB_LOOKUP_SKIP_NEIGH), //L2 NH not resolvable/don't care
+			asm.FnFibLookup.Call(),
+
+			//Recover ifindex to return
+			asm.LoadMem(asm.R0, asm.RFP, -72, asm.Word),
+			asm.Return(),
+		}
+
+		prog, err := ebpf.NewProgram(progSpec)
```
Comment (ti-mo) [inline pkg/datapath/linux/probes/probes.go:505]: This doesn't need to be inside `ns.Do()`, I'd restrict the code inside the closure to what's strictly required (netlink operations and `prog.Test()`).

File: pkg/datapath/linux/probes/probes_test.go (hunk 7 above)
Comment (ti-mo) [reply pkg/datapath/linux/probes/probes_test.go:350]: To add, package features was initially designed to mirror bpftool. Adding this probe there seems a little arbitrary (makingi it hard to draw a line). If the agent needs more bpftool probes, we should put those into the library.

**Review by ti-mo — COMMENTED** [review]
Thanks for picking this up! WIth a bit more polish, this should be good to go. :+1:

File: pkg/datapath/linux/probes/probes.go (hunk 9 above)
Comment (msune) [reply pkg/datapath/linux/probes/probes.go]: Existing probes do the same: https://github.com/cilium/cilium/blob/ab0a12a428cba40e7e662627ceee1c8d5881fd94/pkg/datapath/linux/probes/probes.go#L646

Also, do we care about performance on a probe?

File: pkg/datapath/linux/probes/probes.go (hunk 10 above)
Comment (msune) [reply pkg/datapath/linux/probes/probes.go:459]: It's used in traces
https://github.com/cilium/cilium/blob/ab0a12a428cba40e7e662627ceee1c8d5881fd94/pkg/datapath/linux/probes/probes.go#L506

File: pkg/datapath/linux/probes/probes.go (hunk 11 above)
Comment (msune) [reply pkg/datapath/linux/probes/probes.go:412]: Loopback interfaces are somewhat special in several ways. I wasn't sure if there is a difference in behaviour for `bpf_fib_lookup` for loopbacks. I think a more common use-case for cilium is to attach programs to the vif/netkit interface, so having the probe _as close as the real thing_ as possible

File: pkg/datapath/linux/probes/probes.go (hunk 12 above)
Comment (msune) [reply pkg/datapath/linux/probes/probes.go:505]: For readability or is there a more fundamental reason? Other reviewers didn't mention that.

File: pkg/datapath/linux/probes/probes.go (hunk 9 above)
Comment (ti-mo) [reply pkg/datapath/linux/probes/probes.go]: ~Expressive~, not expensive. Fine by me, just don't like counting zeroes. :stuck_out_tongue:

File: pkg/datapath/linux/probes/probes.go (hunk 10 above)
Comment (ti-mo) [reply pkg/datapath/linux/probes/probes.go:459]: Okay, in that case, shouldn we also put it in the ProgramSpec?

File: pkg/datapath/linux/probes/probes.go (hunk 11 above)
Comment (ti-mo) [reply pkg/datapath/linux/probes/probes.go:412]: Code is liability, so less code is better. If the probe works on a loopback interface, it's preferred over creating extra interfaces.

> I wasn't sure if there is a difference in behaviour for `bpf_fib_lookup` for loopbacks.

Routes added to lo automatically go into the local routing table and get the `host` scope, while routes for other interfaces get the `link` scope by default. I don't think it makes a difference for the probe, but I haven't checked.

> I think a more common use-case for cilium is to attach programs to the vif/netkit interface, so having the probe as close as the real thing as possible.

As close to the real thing would mean creating a veth, not a dummy. Though, for the purpose of a route lookup, this doesn't matter.

File: pkg/datapath/linux/probes/probes.go (hunk 12 above)
Comment (ti-mo) [reply pkg/datapath/linux/probes/probes.go:505]: Sure, readability is an aspect (less indented code), but I think it's a good habit to keep the work done in the other netns to a minimum. Loading bpf programs doesn't depend on being in a particular netns. I struggle to find a good analogy; to me this feels a bit like running things with `sudo` unnecessarily.

It clearly communicates to the reader what's being done to modify the state within the netns. Maybe this is just residual PTSD from working with network namespaces in Go for a decade.

File: pkg/datapath/linux/probes/probes.go (hunk 9 above)
Comment (msune) [reply pkg/datapath/linux/probes/probes.go]: I will adopt the change, and change the other probe.

File: pkg/datapath/linux/probes/probes.go (hunk 11 above)
Comment (msune) [reply pkg/datapath/linux/probes/probes.go:412]: >Routes added to lo automatically go into the local routing table and get the host scope, while routes for other interfaces get the link scope by default. I don't think it makes a difference for the probe, but I haven't checked.

There are many more details (some come from RFCs). See for instance comments like this:

https://github.com/torvalds/linux/blob/master/net/ipv4/route.c#L2544

And others (not extensive):

* https://github.com/torvalds/linux/blob/master/net/ipv4/fib_frontend.c#L389
* https://github.com/torvalds/linux/blob/master/net/ipv4/route.c#L2502
* https://github.com/torvalds/linux/blob/master/net/ipv4/route.c#L2683
* ...

> As close to the real thing would mean creating a veth, not a dummy. Though, for the purpose of a route lookup, this doesn't matter.

I can move it to veth, if you want

File: pkg/datapath/linux/probes/probes.go (hunk 12 above)
Comment (msune) [reply pkg/datapath/linux/probes/probes.go:505]: I don't see any other reason other than readability.

Running inside a NS, a network namespace, has almost no effect in that particular case. If at all, the code is _more_ sandboxed, so it's actually better.

File: pkg/datapath/linux/probes/probes.go (hunk 10 above)
Comment (msune) [reply pkg/datapath/linux/probes/probes.go:459]: >Okay, in that case, shouldn we also put it in the ProgramSpec?

Yes, I will do that

File: pkg/datapath/linux/probes/probes.go (hunk 11 above)
Comment (msune) [reply pkg/datapath/linux/probes/probes.go:412]: Adopted the other 2 changes; I will wait to see if this tests needs to be moved out elsewhere to change it to a veth (albeit dummy and veth are really not that different as opposed to lo vs dummy/veth).

Comment (joestringer) [reply pkg/datapath/linux/probes/probes.go:412]: @ti-mo Maybe this fell off your radar, but for the specific purposes of this PR does it matter to move this to veth or not? If the PR is generally pushing the code into a good direction I don't think we need to make the changes perfect.

File: pkg/datapath/linux/probes/probes.go (hunk 12 above)
Comment (joestringer) [reply pkg/datapath/linux/probes/probes.go:505]: Seems to me like this _hints_ that it's important that the BPF load is run inside the namespace, but since BPF isn't namespaced it misleads the reader a little bit.

I would have a mild preference not to have this run in the namespace similar to Timo but it doesn't seem all that consequential to me.

File: pkg/datapath/linux/probes/probes.go (hunk 11 above)
Comment (joestringer) [reply pkg/datapath/linux/probes/probes.go:412]: > Is the loopback interface actually used for something?

~When I locally deleted all the other namespace iteration / route handling logic, the only dependency I found was to initialize a specific register in the program with an IP address. I don't think the actual IP address matters, so it seems like this probe could be considerably simplified by removing the loopback,route,namespace logic and just putting an RFC5735 address in there. Unless there's something specific about the route add logic that validates the functionality for older kernels? At a glance it just looks like a standard route add.~

EDIT: I see, I mixed up the feedback from Timo below. Reading a bit more the return result from running the program is expected to return the dummy interface's index, so that part is critical to run inside the namespace to find the right route. Disregard the above.

File: pkg/datapath/linux/probes/probes_test.go (hunk 7 above)
Comment (joestringer) [reply pkg/datapath/linux/probes/probes_test.go:350]: @ti-mo @rgo3 @lmb is the feedback here that the probe belongs in cilium/ebpf?
